modules/auxiliary/server/capture/smb.rb

```diff
@@ -25,626 +28,214 @@ def initialize
-          :ntlm_ver => NTLM_CONST::NTLM_V1_RESPONSE,
-          :type => 'ntlm'
-        }
-      )
-        print_status("SMB Capture - NLMv1 Hash correspond to an empty password, ignoring ... #{smb[:ip]}")
-        return
-      end
-      if lm_hash == nt_hash or lm_hash == "" or lm_hash =~ /^0*$/
-        lm_hash_message = "Disabled"
-      elsif NTLM_CRYPT::is_hash_from_empty_pwd?(
-        {
-          :hash => [lm_hash].pack("H*"),
-          :srv_challenge => @challenge,
-          :ntlm_ver => NTLM_CONST::NTLM_V1_RESPONSE,
-          :type => 'lm'
-        }
-      )
-        lm_hash_message = "Disabled (from empty password)"
-      else
-        lm_hash_message = lm_hash
-        lm_chall_message = lm_cli_challenge
-      end
-    when NTLM_CONST::NTLM_V2_RESPONSE
-      if NTLM_CRYPT::is_hash_from_empty_pwd?(
-        {
-          :hash => [nt_hash].pack("H*"),
-          :srv_challenge => @challenge,
-          :cli_challenge => [nt_cli_challenge].pack("H*"),
-          :user => Rex::Text::to_ascii(smb[:username]),
-          :domain => Rex::Text::to_ascii(smb[:domain]),
-          :ntlm_ver => NTLM_CONST::NTLM_V2_RESPONSE,
-          :type => 'ntlm'
-        }
-      )
-        print_status("SMB Capture - NTLMv2 Hash correspond to an empty password, ignoring ... #{smb[:ip]}")
-        return
-      end
+    if datastore['CHALLENGE']
+      if datastore['CHALLENGE'].to_s =~ /^([a-fA-F0-9]{16})$/
+        # Set challenge for all future server responses
 
-      if lm_hash == '0' * 32 and lm_cli_challenge == '0' * 16
-        lm_hash_message = "Disabled"
-        lm_chall_message = 'Disabled'
-      elsif NTLM_CRYPT::is_hash_from_empty_pwd?(
-        {
-          :hash => [lm_hash].pack("H*"),
-          :srv_challenge => @challenge,
-          :cli_challenge => [lm_cli_challenge].pack("H*"),
-          :user => Rex::Text::to_ascii(smb[:username]),
-          :domain => Rex::Text::to_ascii(smb[:domain]),
-          :ntlm_ver => NTLM_CONST::NTLM_V2_RESPONSE,
-          :type => 'lm'
-        }
-      )
-        lm_hash_message = "Disabled (from empty password)"
-        lm_chall_message = 'Disabled'
+        chall = proc { [datastore['CHALLENGE']].pack('H*') }
+        ntlm_provider.generate_server_challenge(&chall)
       else
-        lm_hash_message = lm_hash
-        lm_chall_message = lm_cli_challenge
-      end
-
-    when NTLM_CONST::NTLM_2_SESSION_RESPONSE
-      if NTLM_CRYPT::is_hash_from_empty_pwd?(
-        {
-          :hash => [nt_hash].pack("H*"),
-          :srv_challenge => @challenge,
-          :cli_challenge => [lm_hash].pack("H*")[0,8],
-          :ntlm_ver => NTLM_CONST::NTLM_2_SESSION_RESPONSE,
-          :type => 'ntlm'
-        }
-      )
-        print_status("SMB Capture - NTLM2_session Hash correspond to an empty password, ignoring ... #{smb[:ip]}")
-        return
+        print_warning("#{datastore['CHALLENGE']} is not a valid 16 character hexadecimal pattern, using a random pattern instead.")
```

Comment:
I'm guessing there's no additional code here, because that's the default underlying behavior?
Is it worth defaulting that here explicitly? 🤔

Reply:
I think it would be better to rely on the libraries method of randomization, since it can be trusted to know how to build a good server challenge. Plus if we start hand rolling randomization across multiple modules it won't be consistent and it'll be a pain to update all of them if that's needed in the future.

Reply:
So if the `CHALLENGE` is blank then we should definitely use a random value.

However if the challenge is not blank, and it's invalid because it's not the correct size, then I think we should be failing with a `BadConfig` error. The reason for this would be that it's possible an attacker has a set of Rainbow tables for say `1122334455667788` but they mistype the challenge in as `112233445566778899` or something. Now they're receiving responses with a random challenge and if they missed the warning message, the data they're storing is not going be compatible with their data.